### PR TITLE
latest solc and truffle

### PR DESCRIPTION
**Pull request by johannbarbie**
Resolves https://github.com/leapdao/leap-contracts/issues/110

### CODE REVIEW

File: contracts/ExitHandler.sol [hunk 1]
```diff
@@ -309,16 +309,16 @@ contract ExitHandler is DepositHandler {
         tokens[currentExit.color].addr.transferFrom(address(this), currentExit.owner, currentExit.amount);
       }
       // Pay exit stake
-      currentExit.owner.transfer(currentExit.stake);
+      address(uint160(currentExit.owner)).transfer(currentExit.stake);
```
Comment (troggy) [inline contracts/ExitHandler.sol]: what is going on here? isn't `currentExit.owner` already an `address` ?

File: contracts/FastExitHandler.sol [hunk 2]
```diff
@@ -103,19 +103,21 @@ contract FastExitHandler is ExitHandler {
...
-    uint8 v = uint8(signedData[4]);
-    sigBuff[0] = utxoId;
-    sigBuff[1] = signedData[1];
-    bytes32 sigHash = keccak256(sigBuff);
+    uint8 v = uint8(uint256(signedData[4]));
+    bytes32 sigHash = signedData[1];
+    assembly {
+      mstore(0, utxoId)
+      mstore(0x20, sigHash)
+      sigHash := keccak256(0, 0x40)
```
Comment (troggy) [inline contracts/FastExitHandler.sol]: why this change?

File: contracts/MinGov.sol [hunk 3]
```diff
@@ -69,12 +69,12 @@ contract MinGov is Ownable {
...
           ) {
             // this changes proxy parameters 
-            rv = prop.subject.call(prop.msgData);
+            (rv, ) = prop.subject.call(prop.msgData);
             // use this for 0.5
             // (rv, ) = prop.subject.call(prop.msgData);
           } else {
             // this changes governance parameters to the implementation
-            rv = AdminableProxy(prop.subject).applyProposal(prop.msgData);
+            rv = AdminableProxy(address(uint160(prop.subject))).applyProposal(prop.msgData);
```
Comment (troggy) [inline contracts/MinGov.sol]: same here.. `prop.subject` is an address already, no?

File: contracts/PoaOperator.sol [hunk 4]
```diff
@@ -167,7 +167,12 @@ contract PoaOperator is Adminable {
       // if slot not active, prevent submission
       require(lastCompleteEpoch + 2 < slot.activationEpoch);
     }
-    bytes32 periodRood = keccak256(_blocksRoot, _slotId << 160 | uint160(msg.sender));
+    bytes32 periodRood = bytes32(_slotId << 160 | uint160(msg.sender));
```
Comment (troggy) [inline contracts/PoaOperator.sol]: replace with `keccak256(abi.encodePacked(_blocksRoot, _slotId << 160 | uint160(msg.sender)))` instead of using assembly?

File: contracts/TxLib.sol [hunk 5]
```diff
@@ -56,7 +55,7 @@ library TxLib {
         // load the depositId (4 bytes) starting from byte 2 of tx
         inputData := mload(add(add(offset, 4), _txData))
       }
-      inputData = bytes32(uint32(inputData));
+      inputData = bytes32(uint256(uint32(uint256(inputData))));
```
Comment (troggy) [inline contracts/TxLib.sol]: image

File: package.json [hunk 6]
```diff
@@ -2,6 +2,7 @@
   "name": "leap-contracts",
   "version": "0.1.0",
   "description": "LeapDAO bridge contracts (Plasma + PoS)",
+  "main": "truffle-config.js",
```
Comment (troggy) [inline package.json]: remove this

**Comment by troggy** [comment]
also let's finally choose what we use: `yarn` or `npm` and remove either `package-lock.json` or `yarn.lock`. I suggest to stick with `yarn` and drop `package-lock.json` leaving `yarn.lock`

File: contracts/ExitHandler.sol (hunk 1 above)
Comment (johannbarbie) [reply contracts/ExitHandler.sol]: there is a difference between `address` and `address payable` now, only second one can be transfered to. in can only be cast from `uint160`. 🤷‍♂️

File: contracts/FastExitHandler.sol (hunk 2 above)
Comment (johannbarbie) [reply contracts/FastExitHandler.sol]: because it would have required `abi.encoded`, which is not gas efficient.

File: contracts/PoaOperator.sol (hunk 4 above)
Comment (johannbarbie) [reply contracts/PoaOperator.sol]: abi.encoded is super gas inefficient. to be avoided

Comment (johannbarbie) [reply contracts/PoaOperator.sol]: https://twitter.com/nicksdjohnson/status/1044904764146962432

File: contracts/FastExitHandler.sol (hunk 2 above)
Comment (johannbarbie) [reply contracts/FastExitHandler.sol]: https://twitter.com/nicksdjohnson/status/1044904764146962432

File: contracts/MinGov.sol (hunk 3 above)
Comment (johannbarbie) [reply contracts/MinGov.sol]: there is a difference between address and address payable now, only second one can be transfered to. in can only be cast from uint160. 🤷‍♂️

**Comment by troggy** [comment]
approved, except Travis is broken. Probably `yarn.lock` is outdated?

**Comment by johannbarbie** [comment]
> Probably yarn.lock is outdated?
I made it fresh

**Comment by johannbarbie** [comment]
the `main` in package.json was actually needed.
